### ISSUE TITLE
Large but empty delete queue ZAP causes slow mount

### ISSUE THREAD

**Issue by suuchan** [openzfs/zfs#12006]
### System information

Type | Version/Name
 --- | ---
Distribution Name	| RHEL
Distribution Version	| 8.3
Linux Kernel	| 4.18.0-240.22.1.el8_3.x86_64
Architecture	| x86_64
ZFS Version	| 2.0.4-1
SPL Version	| 2.0.4-1

We have exactly the same problem as described in issue #5916 which is now 4 years old and as far as I can see unresolved.

One file system with 47TB customer data has ended up with a 7GB delete queue. As a result, a mount after import takes over 30 minutes to complete. The delete queue does not go away with send/receive.  After deleting all files on a clone, scrubbing, re-importing, etc. the ZAP remains and `zdb -dddd` also takes 30 minutes to report this:

```
...
	Fat ZAP stats:
		Pointer table:
			524288 elements
			zt_blk: 248568
			zt_numblks: 256
			zt_shift: 19
			zt_blks_copied: 0
			zt_nextblk: 0
		ZAP entries: 0
		Leaf blocks: 445397
		Total blocks: 445909
		zap_block_type: 0x8000000000000001
		zap_magic: 0x2f52ab2ab
		zap_salt: 0x2d0d489983
		Leafs with 2^n pointers:
			  0: 376558 ****************************************
			  1:  63813 *******
			  2:   5026 *
		Blocks with n*5 entries:
			  0: 445397 ****************************************
		Blocks n/10 full:
			  1: 445397 ****************************************
		Entries with n chunks:
		Buckets with n entries:
			  0: 228043264 ****************************************
...
```

Is it still impossible to remove/shrink a huge ZAP or otherwise tweak the mount process to just ignore it if it is empty?

**Comment by ahrens** [comment]
That's correct.  For this special case, it probably wouldn't be too hard to destroy & recreate the delete queue ZAP object if it has zero entries but more than a handful of blocks.

**Comment by suuchan** [comment]
Our primary issue is the long mount time.  This issue occurs on a file server with data of many customers and exports can only be resumed when all file systems are mounted. So this one defect affects many people in our case and we basically cannot reboot this server for maintenance without downtime, even though it is a HA setup.

Any solution that fixes this issue would be very welcome. If there is a way to just de-reference the ZAP, even if the space is forever lost, that would also work for our case. Leaving the ZAP but ignoring it is it is clearly empty sounds like the easiest workaround to me. The mount would still be slow if it is not empty but as I understand it that would not be the case of a regular umount->export->import->mount cycle.

**Comment by allanjude** [comment]
KlaraSystems.com has done some work similar to this in the past, around dealing with DDT ZAPs that are too large, or sparsely populated like this.

If you are interested in talking to us about solving this issue, please get in touch. allan@klarasystems.com

**Comment by GregorKopka** [comment]
Why is the stale bot still operating on issues marked as 'defect' ?

**Comment by behlendorf** [comment]
@GregorKopka since most issues are marked as defect and it gives us a chance to revisit them.  This is in fact still an issue until the ZAP code is extended to support shrinking.

**Comment by jumbi77** [comment]
The ZAP shrinking feature was shortly discussed recently on OpenZFS Leadership Meeting [here](https://youtu.be/GPgy9ATYeYY?t=2635).

Referencing the [illumos PR](https://www.illumos.org/issues/13916#change-34349) and corresponding [PR on Gerrit](https://code.illumos.org/c/illumos-gate/+/1580).

Also referencing an older OpenZFS issue, which would also benefit from ZAP shrinking possibility [#4933](https://github.com/openzfs/zfs/issues/4933).

**Comment by jumbi77** [comment]
The ZAP shrinking feature was shortly discussed recently on OpenZFS Leadership Meeting [here](https://youtu.be/GPgy9ATYeYY?t=2635).

Referencing the [illumos PR](https://www.illumos.org/issues/13916#change-34349) and corresponding [PR on Gerrit](https://code.illumos.org/c/illumos-gate/+/1580).

Also referencing an older OpenZFS issue, which would also benefit from ZAP shrinking possibility [#4933](https://github.com/openzfs/zfs/issues/4933).

Last but not least referencing this [discussion](https://github.com/openzfs/zfs/discussions/8420).

**Comment by HPPinata** [comment]
With https://github.com/openzfs/zfs/pull/15888 merged isn't this solved by now and can be closed?
